### PR TITLE
Keep 'live' tab shown while uploading for improved feedback

### PR DESCRIPTION
**Pull request by Martchus**
See https://progress.opensuse.org/issues/56591 and particular commit messages

### CODE REVIEW

**Comment by Martchus** [comment]
I was really expecting this could be achieved easier. The adjustments in the worker code at least don't seem to break everything but I have to think of a better idea to test this. Maybe I just do it manually after all but fake some chunk upload progress. Sounds at least better than introducing an unstable automatic test or something with insanely complex mocking. Likely some unit tests will break anyways so this change wouldn't be without any reflection within the test suite.

**Comment by Martchus** [comment]
That's how it looks like (with some fake data because I didn't always want to wait for a job which creates big artifacts):  
screenshot_20200526_171711

So the developer mode controls and the screen are hidden when the job state turns into uploading and the worker keeps updating the live log while uploading big files such as HDD images.

File: lib/OpenQA/Worker/Job.pm [hunk 1]
```diff
@@ -75,22 +75,23 @@ sub new {
...
+    $self->{_images_to_send}               = {};
+    $self->{_files_to_send}                = {};
+    $self->{_known_images}                 = [];
+    $self->{_known_files}                  = [];
+    $self->{_last_screenshot}              = '';
+    $self->{_current_test_module}          = undef;
+    $self->{_progress_info}                = {};
+    $self->{_engine}                       = undef;
+    $self->{_is_uploading_results}         = 0;
+    $self->{_has_uploaded_logs_and_assets} = 0;
```
Comment (Martchus) [inline lib/OpenQA/Worker/Job.pm:94]: I only added `_has_uploaded_logs_and_assets`. All the other lines are just tidy.

**Comment by Martchus** [comment]
I've used my test runs for https://progress.opensuse.org/issues/65004 to test this PR as well and found one issue which now has been fixed.

I was too slow to make a screenshot but that's how the real log data looks like:

```
[1;94m[2020-05-27T17:04:36.0227 CEST] [info] [pid:31306] Uploading logs_from_installation_system-y2logs.tar.bz2[0m
[1;94m[2020-05-27T17:04:36.0256 CEST] [info] [pid:31306] Uploading SLES-15-SP1-x86_64-mru-install-minimal-with-addons-Build:15220:susefirewall2-to-firewalld-Server-DVD-Incidents-64bit.qcow2[0m
[1;94m[2020-05-27T17:04:43.0640 CEST] [info] [pid:31306] SLES-15-SP1-x86_64-mru-install-minimal-with-addons-Build:15220:susefirewall2-to-firewalld-Server-DVD-Incidents-64bit.qcow2: Processing chunk 10/1176, avg. speed ~976.562 KiB/s[0m
[1;94m[2020-05-27T17:04:44.0302 CEST] [info] [pid:31306] SLES-15-SP1-x86_64-mru-install-minimal-with-addons-Build:15220:susefirewall2-to-firewalld-Server-DVD-Incidents-64bit.qcow2: Processing chunk 20/1176, avg. speed ~976.562 KiB/s[0m
[1;94m[2020-05-27T17:04:44.0942 CEST] [info] [pid:31306] SLES-15-SP1-x86_64-mru-install-minimal-with-addons-Build:15220:susefirewall2-to-firewalld-Server-DVD-Incidents-64bit.qcow2: Processing chunk 30/1176, avg. speed ~976.562 KiB/s[0m
[1;94m[2020-05-27T17:04:45.0636 CEST] [info] [pid:31306] SLES-15-SP1-x86_64-mru-install-minimal-with-addons-Build:15220:susefirewall2-to-firewalld-Server-DVD-Incidents-64bit.qcow2: Processing chunk 40/1176, avg. speed ~976.562 KiB/s[0m
…
[1;94m[2020-05-27T17:06:05.0399 CEST] [info] [pid:31306] SLES-15-SP1-x86_64-mru-install-minimal-with-addons-Build:15220:susefirewall2-to-firewalld-Server-DVD-Incidents-64bit.qcow2: Processing chunk 1170/1176, avg. speed ~976.562 KiB/s[0m
[1;94m[2020-05-27T17:06:10.0968 CEST] [info] [pid:31306] SLES-15-SP1-x86_64-mru-install-minimal-with-addons-Build:15220:susefirewall2-to-firewalld-Server-DVD-Incidents-64bit.qcow2: Processing chunk 1176/1176, avg. speed ~139.812 KiB/s[0m
[1;94m[2020-05-27T17:06:10.0970 CEST] [info] [pid:31306] Uploading video.ogv[0m
[1;94m[2020-05-27T17:06:11.0068 CEST] [info] [pid:31306] Uploading video_time.vtt[0m
[1;94m[2020-05-27T17:06:11.0082 CEST] [info] [pid:31306] Uploading vars.json[0m
[1;94m[2020-05-27T17:06:11.0095 CEST] [info] [pid:31306] Uploading serial0.txt[0m
[1;94m[2020-05-27T17:06:11.0112 CEST] [info] [pid:31306] Uploading autoinst-log.txt[0m
```

Maybe we should at least in the JavaScript client strip the escape characters.

**Comment by Amrysliu** [comment]
I did some tests in my local. In my test, the `test_order.json` was not generated (It was an accident), but the job was not finished, its status was `assigned` or `uploading` (one time was `assigned`, and more than one time was `uploading`). I also did same test using `master` branch code, the job would be set finished, its status was `incomplete`.
Here is the worker log use `master` branch: 

```
[debug] [pid:10394] Running job 642 from http://127.0.0.1:9526: 00000642-sle-15-SP2-Online-x86_64-Build0368-test3@64bit.
[debug] [pid:10394] Unable to read test_order.json: No such file or directory
[debug] [pid:10394] Stopping job 642 from http://127.0.0.1:9526: 00000642-sle-15-SP2-Online-x86_64-Build0368-test3@64bit - reason: no tests scheduled
[debug] [pid:10394] REST-API call: POST http://127.0.0.1:9526/api/v1/jobs/642/status
[info] [pid:10394] Trying to stop job gracefully by announcing it to command server via http://localhost:20013/XYzix0InZOZ4ngYo/broadcast
[info] [pid:10394] Unable to stop the command server gracefully: 
[info] [pid:10394] Command server likely not reachable at all
[info] [pid:10394] Isotovideo exit status: 1
[info] [pid:10394] +++ worker notes +++
[info] [pid:10394] End time: 2020-05-29 03:58:52
[info] [pid:10394] Result: no tests scheduled
[info] [pid:10443] Uploading video.ogv
[info] [pid:10443] Uploading video_time.vtt
[info] [pid:10443] Uploading vars.json
[info] [pid:10443] Uploading serial0.txt
[info] [pid:10443] Uploading autoinst-log.txt
[info] [pid:10443] Uploading serial_terminal.txt
[info] [pid:10443] Uploading worker-log.txt
[debug] [pid:10394] Job 642 stopped as incomplete
[debug] [pid:10394] Unable to read test_order.json: No such file or directory
[debug] [pid:10394] REST-API call: POST **http://127.0.0.1:9526/api/v1/jobs/642/set_done?reason=no+tests+scheduled&result=incomplete**
[debug] [pid:10394] Job 642 from http://127.0.0.1:9526 finished - reason: no tests scheduled
```
Here is the worker log using this pr

```
[debug] [pid:10835] Running job 643 from http://127.0.0.1:9526: 00000643-sle-15-SP2-Online-x86_64-Build0368-test3@64bit.
[debug] [pid:10835] Unable to read test_order.json: No such file or directory
[debug] [pid:10835] Stopping job 643 from http://127.0.0.1:9526: 00000643-sle-15-SP2-Online-x86_64-Build0368-test3@64bit - reason: no tests scheduled
[debug] [pid:10835] REST-API call: POST http://127.0.0.1:9526/api/v1/jobs/643/status
[error] [pid:10835] REST-API error (POST http://127.0.0.1:9526/api/v1/jobs/643/status): 400 response: Got status update for job 643 and worker 1 but there is not even a worker assigned to this job (job is scheduled) (remaining tries: 0)
[info] [pid:10835] Trying to stop job gracefully by announcing it to command server via http://localhost:20013/WpbpIAiMHj0TPuEm/broadcast
[info] [pid:10835] Unable to stop the command server gracefully: 
[info] [pid:10835] Command server likely not reachable at all
[info] [pid:10835] Isotovideo exit status: 1
[info] [pid:10835] +++ worker notes +++
[info] [pid:10835] End time: 2020-05-29 04:00:45
[info] [pid:10835] Result: no tests scheduled
[info] [pid:10857] Uploading video.ogv
[info] [pid:10857] Uploading video_time.vtt
[info] [pid:10857] Uploading vars.json
[info] [pid:10857] Uploading serial0.txt
[info] [pid:10857] Uploading autoinst-log.txt
[info] [pid:10857] Uploading serial_terminal.txt
[info] [pid:10857] Uploading worker-log.txt
```
There is no `set_done` operation.

I am not sure if the test is correct, please confirm that. Thanks.

File: lib/OpenQA/Worker/Job.pm (hunk 1 above)
Comment (kalikiana) [reply lib/OpenQA/Worker/Job.pm:94]: I wish there was a way to make these less annoying to review... I end up counting the lines and looking for the new one...

**Comment by Martchus** [comment]
@Amrysliu Are you saying that `test_order.json` was not generated because of something unrelated to that PR but this PR made handling this situation worse because the job was not marked as done anymore? I belief I have tested this case locally when breaking os-autoinst during the development of the video encoder changes but I can check it again, e.g. by creating a syntax error in the test suite.

**Comment by Amrysliu** [comment]
> @Amrysliu Are you saying that `test_order.json` was not generated because of something unrelated to that PR but this PR made handling this situation worse because the job was not marked as done anymore? I belief I have tested this case locally when breaking os-autoinst during the development of the video encoder changes but I can check it again, e.g. by creating a syntax error in the test suite.

yes, I mean that. Thanks for checking again.

**Comment by Martchus** [comment]
I couldn't reproduce the problem locally. Here it looks like this:

```
[debug] [pid:15704] Stopping job 1381 from http://localhost:9526: 00001381-opensuse-Tumbleweed-DVD-x86_64-Build20200523-boot_to_snapshot@64bit - reason: died
[debug] [pid:15704] REST-API call: POST http://localhost:9526/api/v1/jobs/1381/status
[info] [pid:15704] +++ worker notes +++
[info] [pid:15704] End time: 2020-05-29 10:05:40
[info] [pid:15704] Result: died
[info] [pid:16350] Uploading vars.json
[info] [pid:16350] Uploading autoinst-log.txt
[info] [pid:16350] Uploading worker-log.txt
[debug] [pid:15704] Job 1381 stopped as incomplete
[debug] [pid:15704] Unable to read test_order.json: No such file or directory
[debug] [pid:15704] REST-API call: POST http://localhost:9526/api/v1/jobs/1381/set_done?reason=tests+died%3A+unable+to+load+main.pm%2C+check+the+log+for+the+cause+%28e.g.+syntax+error%29&result=incomplete
[debug] [pid:15704] Job 1381 from http://localhost:9526 finished - reason: died
[debug] [pid:15704] Cleaning up for next job
[warn] [pid:15704] Ignoring WS message from http://localhost:9526 with type livelog_stop and job ID 1381 (currently not executing a job):
{ jobid => 1381, type => "livelog_stop" }
```

But I've noticed another problem: It tries to stop the live log at a point when it is not possible/useful anymore. However, I guess that this problem was technically possible before as well and it is now just more likely to happen. The good thing is that it is absolutely not breaking anything - it only disturbs the log.

**Comment by kalikiana** [comment]
FYI I took care of the warning in `t/24-worker-jobs.t` in #3130 - I noticed it in this PR but it's not being introduced here.

**Comment by Amrysliu** [comment]
@Martchus Thanks for checking.  Seems there is something wrong in my environment.

**Pull request merged**